### PR TITLE
Cleanup

### PR DESCRIPTION
**Pull request by riamon-v**
Advanced on #95, changes Server into a Singleton, with it a reduction of the pointer amount but there may still be some changes to be made.
Let's review !

### CODE REVIEW

File: include/Keyboard.hpp [hunk 1]
```diff
@@ -29,7 +29,8 @@ class Keyboard : public KeyboardListeners
   bool debug = false;
 
 private:
-  Server *server;
+  Server &server;
```
Comment (Gaspard--) [inline include/Keyboard.hpp]: Didn't server become a singleton?

File: include/Output.hpp [hunk 2]
```diff
@@ -41,7 +41,7 @@ class Output : public OutputListeners
   wlr_box saved;
 
 private:
-  Server *server;
+  Server &server;
```
Comment (Gaspard--) [inline include/Output.hpp]: See above.

File: include/Seat.hpp [hunk 3]
```diff
@@ -13,13 +13,13 @@ struct SeatListeners
...
-  Seat(Server *server);
+  Seat();
   ~Seat() = default;
 
   void seat_request_cursor(struct wl_listener *listener, void *data);
   struct wlr_seat *getSeat() const noexcept;
 
 private:
-  Server *server;
+  Server &server;
```
Comment (Gaspard--) [inline include/Seat.hpp]: Same as above.

File: include/ServerCursor.hpp [hunk 4]
```diff
@@ -39,7 +39,7 @@ class ServerCursor : public ServerCursorListeners
   CursorMode cursor_mode;
 
 private:
-  Server *server;
+  Server &server;
```
Comment (Gaspard--) [inline include/ServerCursor.hpp]: See above

File: include/ServerInput.hpp [hunk 5]
```diff
@@ -17,13 +17,13 @@ struct ServerInputListeners
...
 public:
-  ServerInput(Server *server);
+  ServerInput();
   ~ServerInput() = default;
 
   void server_new_input(struct wl_listener *listener, void *data);
 
 private:
-  Server *server;
+  Server &server;
```
Comment (Gaspard--) [inline include/ServerInput.hpp]: See above

File: include/ServerOutput.hpp [hunk 6]
```diff
@@ -40,7 +40,7 @@ class ServerOutput : public ServerOutputListeners
   Output const &getOutput(wlr_output *wlr_output) const noexcept;
 
 private:
-  Server *server;
+  Server &server;
```
Comment (Gaspard--) [inline include/ServerOutput.hpp]: See above

File: include/View.hpp [hunk 7]
```diff
@@ -39,12 +39,12 @@ class View : public ViewListeners
...
   void focus_view();
 
-  static View *desktop_view_at(Server *server, double lx, double ly,
+  static View *desktop_view_at(Server &server, double lx, double ly,
 			struct wlr_surface **surface, double *sx, double *sy);
 
   struct wlr_output *getOutput();
 
-  Server *server;
+  Server &server;
```
Comment (Gaspard--) [inline include/View.hpp]: See above

File: include/XdgShell.hpp [hunk 8]
```diff
@@ -15,14 +15,14 @@ struct XdgShellListeners
...
+  XdgShell();
   ~XdgShell() = default;
 
   void xdg_surface_destroy(struct wl_listener *listener, void *data);
   void server_new_xdg_surface(struct wl_listener *listener, void *data);
   void xdg_handle_new_popup(struct wl_listener *listenr, void *data);
 
 private:
-  Server *server;
+  Server &server;
```
Comment (Gaspard--) [inline include/XdgShell.hpp]: See above

**Review by Gaspard-- — CHANGES_REQUESTED** [review]

File: include/Server.hpp [hunk 9]
```diff
@@ -37,6 +37,8 @@ class Server
     }
   };
 
+  static Server &getInstance();
```
Comment (Gaspard--) [inline include/Server.hpp]:
```suggestion
  static Server &getInstance() noexcept;
```
(need to update the implementation prototype in the .cpp too ofc.)

File: include/View.hpp [hunk 10]
```diff
@@ -39,12 +39,12 @@ class View : public ViewListeners
   void close();
   void focus_view();
 
-  static View *desktop_view_at(Server *server, double lx, double ly,
+  static View *desktop_view_at(Server &server, double lx, double ly,
```
Comment (Gaspard--) [inline include/View.hpp]:
```suggestion
  static View *desktop_view_at(double lx, double ly,
```
(update in .cpp too)

File: source/Commands.cpp [hunk 11]
```diff
@@ -4,9 +4,11 @@
 
 namespace
 {
-  auto getContainerFocusedView(Server *server, wm::WindowTree &windowTree, wm::WindowNodeIndex containerNode, wm::WindowNodeIndex viewNode)
+  auto getContainerFocusedView(wm::WindowTree &windowTree, wm::WindowNodeIndex containerNode, wm::WindowNodeIndex viewNode)
   {
-    for (auto &tmpView: server->views)
+    Server &server = Server::getInstance();
+
+    for (auto &tmpView: server.views)
```
Comment (Gaspard--) [inline source/Commands.cpp]:
```suggestion
    for (auto &tmpView : server.views)
```

File: include/Server.hpp [hunk 12]
```diff
@@ -21,9 +21,9 @@ enum class OpenType : uint8_t
 class Server
 {
 public:
-  Server();
   ~Server();
```
Comment (Gaspard--) [inline include/Server.hpp:26]:
```suggestion
  Server(Server const &) = delete;
  Server(Server &&) = delete;  
  ~Server();
```

**Review by Gaspard-- — CHANGES_REQUESTED** [review]

File: include/View.hpp (hunk 10 above)
Comment (riamon-v) [reply include/View.hpp]: This can't work while the function is static, I already tried it

Comment (Gaspard--) [reply include/View.hpp]: Why? You just need to make a getInstance on the server singleton right?

File: source/Server.cpp [hunk 13]
```diff
@@ -34,6 +36,11 @@ Server::~Server()
   // wl_display_destroy(display);
 }
 
+Server &Server::getInstance() noexcept
```
Comment (Gaspard--) [inline source/Server.cpp]: Not a huge deal, but I'd probably move this in the .hpp. Easier inlining for the compiler, which means either improved performance or link time.

File: source/View.cpp [hunk 14]
```diff
@@ -266,10 +270,11 @@ bool View::at(double lx, double ly, struct wlr_surface **surface, double *sx, do
   return false;
 }
 
-View *View::desktop_view_at(Server *server, double lx, double ly,
+View *View::desktop_view_at(double lx, double ly,
 			    struct wlr_surface **surface, double *sx, double *sy)
 {
-  for (auto &view : server->views)
+  Server &server  = Server::getInstance();
```
Comment (Gaspard--) [inline source/View.cpp]:
```suggestion
  Server &server = Server::getInstance();
```

**Review by Gaspard-- — CHANGES_REQUESTED** [review]

**Comment by Gaspard--** [comment]
(just fixed the non-compiling macro)

File: include/Server.hpp [hunk 15]
```diff
@@ -37,6 +39,10 @@ class Server
     }
   };
 
+  static Server &getInstance() noexcept {
```
Comment (Gaspard--) [inline include/Server.hpp]:
```suggestion
  static Server &getInstance() noexcept
  {
```

**Review by Gaspard-- — APPROVED** [review]

**Pull request merged**